### PR TITLE
Fix #1881

### LINKED ISSUE #1881: BufferedInStream does not read all of stdin if some of the input is delayed
`a.zig`
```rust
const std = @import("std");

pub fn main() !void {
    const stdout = &(try std.io.getStdOut()).outStream().stream;

    try stdout.write("Test 1\n");

    var i: usize = 0;
    while (i < 100000000) : (i += 1) {}

    try stdout.write("Test 2\n");
}
```

`b.zig`
```rust
const std = @import("std");

const debug = std.debug;
const io = std.io;
const heap = std.heap;
const os = std.os;

const BufInStream = io.BufferedInStream(os.File.InStream.Error);

pub fn main() !void {
    const stdin = &BufInStream.init(&(try std.io.getStdIn()).inStream().stream).stream;

    var direct_allocator = std.heap.DirectAllocator.init();
    defer direct_allocator.deinit();

    var arena = heap.ArenaAllocator.init(&direct_allocator.allocator);
    defer arena.deinit();

    const allocator = &arena.allocator;

    while (stdin.readUntilDelimiterAlloc(allocator, '\n', 10000)) |line| {
        debug.warn("{}\n", line);
    } else |err| {}
}
```

```
$ ./a | ./b
Test 1
```

### PR DESCRIPTION
**Pull request by sjdh02**
Should fix #1881.

### CODE REVIEW

**Comment by kristate** [comment]
Thanks for the patch. Do you have a test case for this?

**Comment by Hejsil** [comment]
Here is a self contained test case. @sjdh02 could you add this?
```rust
const std = @import("std");

const debug = std.debug;
const heap = std.heap;
const io = std.io;
const mem = std.mem;

const OneByteReadInStream = struct {
    const Error = error{NoError};
    const Stream = io.InStream(Error);

    stream: Stream,
    str: []const u8,
    curr: usize,

    fn init(str: []const u8) OneByteReadInStream {
        return OneByteReadInStream{
            .stream = Stream{ .readFn = readFn },
            .str = str,
            .curr = 0,
        };
    }

    fn readFn(in_stream: *Stream, dest: []u8) Error!usize {
        const self = @fieldParentPtr(OneByteReadInStream, "stream", in_stream);
        if (self.str.len <= self.curr)
            return 0;
        if (dest.len == 0)
            return 0;

        dest[0] = self.str[self.curr];
        self.curr += 1;
        return 1;
    }
};

test "" {
    var buf: [100]u8 = undefined;
    const allocator = &heap.FixedBufferAllocator.init(buf[0..]).allocator;

    const str = "This is a test";
    var one_byte_stream = OneByteReadInStream.init(str);
    var buf_in_stream = io.BufferedInStream(OneByteReadInStream.Error).init(&one_byte_stream.stream);
    const stream = &buf_in_stream.stream;

    const res = try stream.readAllAlloc(allocator, str.len + 1);
    debug.assertOrPanic(mem.eql(u8, str, res));
}
```

**Comment by sjdh02** [comment]
@Hejsil Done. Let me know if you need me to move it somewhere else, I wasn't really sure where to put it so I just picked a seemingly reasonable place.

File: test/cases/bufferedinstream.zig [hunk 1]
```diff
@@ -0,0 +1,48 @@
+const std = @import("std");
```
Comment (Hejsil) [inline test/cases/bufferedinstream.zig]: We probably want this test in `io.zig`

File: std/io.zig [hunk 2]
```diff
@@ -340,7 +340,8 @@ pub fn BufferedInStreamCustom(comptime buffer_size: usize, comptime Error: type)
                 const amt_buffered = self.end_index - self.start_index;
                 if (amt_buffered == 0) {
                     assert(self.end_index <= buffer_size);
-                    if (self.end_index == buffer_size) {
+                    // Make sure the last read actually gave us some data
```
Comment (kristate) [inline std/io.zig]: This is better:
```suggestion
                    // Make sure the last read actually gave us some data
                    if (self.end_index == 0) {
                        // reading from the unbuffered stream returned nothing
                        // so we have nothing left to read.
                        return dest_index;
                    }
                    // we can read more data from the unbuffered stream
                    if (dest_space < buffer_size) {
                        self.start_index = 0;
                        self.end_index = try self.unbuffered_in_stream.read(self.buffer[0..]);
                    } else {
                        // asking for so much data that buffering is actually less efficient.
                        // forward the request directly to the unbuffered stream
                        const amt_read = try self.unbuffered_in_stream.read(dest[dest_index..]);
                        return dest_index + amt_read;
                    }
```

**Comment by sjdh02** [comment]
Alright, I've taken care of both of your changes, thanks for the feedback.

I had to have the test case call `@This()` instead of just referencing the struct by name, for, err, some reason. I don't know enough about how zig sees test blocks to actually explain it, but it appears to work fine.

**Comment by Hejsil** [comment]
@sjdh02 In global scope there is no order of names, but in local scopes, order matters. `OneByteReadInStream` has not been defined when `init` was declared. `@This()` will return the `type` you're currently in, which allows you to refer to `OneByteReadInStream` before it was bound to its name. I hope that makes sense :)

**Comment by sjdh02** [comment]
I see, that makes sense. Thanks for the info.

**Comment by andrewrk** [comment]
@sjdh02 can you squash your commits and rebase against master? sorry about the conflicts.

**Comment by sjdh02** [comment]
Done, thanks for reminding me to check up on this.

**Comment by andrewrk** [comment]
Thanks!

**Pull request merged**